File: apps/backend/kubedeck_backend/api/secrets.py
from __future__ import annotations

import base64
import binascii
from typing import Any

from pydantic import BaseModel

from kubedeck_backend.api.runtime import cluster_command, runner, store
from kubedeck_backend.api.validation import api_error, validate_identifier
from kubedeck_backend.core.audit import append_audit_event
from kubedeck_backend.kubectl.command import KubectlError

SECRET_JSON_MAX_OUTPUT_BYTES = 8 * 1024 * 1024
SECRET_VALUE_MAX_BYTES = 2 * 1024 * 1024


class SecretRevealRequest(BaseModel):
    key: str


class SecretCopyAuditRequest(BaseModel):
    key: str


def secret_reveal_timeout_seconds() -> int:
    try:
        return max(1, min(300, int(store.load().settings.secretRevealTimeoutSeconds)))
    except Exception:
        return 30


def load_secret_raw(cluster_id: str, namespace: str, name: str) -> dict[str, Any]:
    command = cluster_command(
        cluster_id,
        ["get", "secret", name, "-n", namespace, "-o", "json"],
        timeout=30,
        max_output_bytes=SECRET_JSON_MAX_OUTPUT_BYTES,
    )
    payload = runner.run_json(command)
    if not isinstance(payload, dict):
        raise api_error(500, "INVALID_SECRET_PAYLOAD", "kubectl returned an invalid Secret payload", command_preview=command.preview)
    return payload


def secret_keys_response(cluster_id: str, namespace: str, name: str) -> dict[str, Any]:
    secret = load_secret_raw(cluster_id, namespace, name)
    data = secret_data_map(secret)
    keys: list[dict[str, Any]] = []
    for key in sorted(data):
        encoded = data.get(key) or ""
        encoded_bytes = len(encoded.encode("utf-8", "replace"))
        decoded_bytes = 0
        valid_base64 = True
        binary = False
        try:
            decoded = base64.b64decode(encoded.encode("utf-8"), validate=True)
            decoded_bytes = len(decoded)
            binary = is_binary_payload(decoded)
        except (binascii.Error, ValueError):
            valid_base64 = False
        keys.append({
            "key": key,
            "encodedBytes": encoded_bytes,
            "decodedBytes": decoded_bytes,
            "validBase64": valid_base64,
            "binary": binary,
        })

    metadata = secret.get("metadata") or {}
    return {
        "type": str(secret.get("type") or "Opaque"),
        "immutable": bool(secret.get("immutable", False)),
        "namespace": str(metadata.get("namespace") or namespace),
        "name": str(metadata.get("name") or name),
        "keys": keys,
        "revealTimeoutSeconds": secret_reveal_timeout_seconds(),
    }


def reveal_secret_key(cluster_id: str, namespace: str, name: str, key: str) -> dict[str, Any]:
    safe_key = validate_secret_key(key)
    secret = load_secret_raw(cluster_id, namespace, name)
    data = secret_data_map(secret)
    if safe_key not in data:
        raise api_error(404, "SECRET_KEY_NOT_FOUND", f"Secret key was not found: {safe_key}")

    encoded = data.get(safe_key) or ""
    try:
        decoded = base64.b64decode(encoded.encode("utf-8"), validate=True)
    except (binascii.Error, ValueError):
        append_audit_event(
            action="secret.reveal",
            status="failed",
            cluster_id=cluster_id,
            namespace=namespace,
            resource="secrets",
            name=name,
            message="invalid base64 data",
            extra={"key": safe_key},
        )
        raise api_error(400, "SECRET_VALUE_INVALID_BASE64", f"Secret key is not valid base64 data: {safe_key}")

    if len(decoded) > SECRET_VALUE_MAX_BYTES:
        append_audit_event(
            action="secret.reveal",
            status="failed",
            cluster_id=cluster_id,
            namespace=namespace,
            resource="secrets",
            name=name,
            message="secret value too large to reveal",
            extra={"key": safe_key, "decodedBytes": len(decoded)},
        )
        raise api_error(413, "SECRET_VALUE_TOO_LARGE", f"Secret value is too large to reveal safely ({len(decoded)} bytes)")

    value = decoded.decode("utf-8", "replace")
    binary = is_binary_payload(decoded)
    append_audit_event(
        action="secret.reveal",
        status="success",
        cluster_id=cluster_id,
        namespace=namespace,
        resource="secrets",
        name=name,
        extra={"key": safe_key, "decodedBytes": len(decoded), "binary": binary},
    )
    return {
        "key": safe_key,
        "value": value,
        "decodedBytes": len(decoded),
        "binary": binary,
        "revealTimeoutSeconds": secret_reveal_timeout_seconds(),
    }


def audit_secret_copy(cluster_id: str, namespace: str, name: str, key: str) -> dict[str, bool]:
    safe_key = validate_secret_key(key)
    append_audit_event(
        action="secret.copy",
        status="success",
        cluster_id=cluster_id,
        namespace=namespace,
        resource="secrets",
        name=name,
        extra={"key": safe_key},
    )
    return {"ok": True}


def secret_data_map(secret: dict[str, Any]) -> dict[str, str]:
    data = secret.get("data") or {}
    if not isinstance(data, dict):
        return {}
    return {str(key): str(value) for key, value in data.items() if key is not None}


def validate_secret_key(value: str) -> str:
    return validate_identifier(value, "secret key", max_length=512)


def is_binary_payload(value: bytes) -> bool:
    if not value:
        return False
    if b"\x00" in value:
        return True
    text_bytes = sum(1 for byte in value if byte in b"\n\r\t" or 32 <= byte <= 126)
    return (text_bytes / len(value)) < 0.85

File: apps/desktop/src/main/backend/routes/secrets.ts
import type { IncomingMessage, ServerResponse } from "node:http";
import type { AuditStore } from "../audit/auditStore";
import { ClusterNotFoundError, type ConfigStore } from "../config/configStore";
import { writeError } from "../errors";
import { readJsonBody, RequestBodyError, writeJson } from "../http";
import { clusterCommand } from "../kubectl/clusterCommand";
import { KubectlError, writeKubectlError } from "../kubectl/errors";
import type { KubectlRunner } from "../kubectl/runner";
import { RequestValidationError, validateIdentifier } from "../validation";

const SECRET_JSON_MAX_OUTPUT_BYTES = 8 * 1024 * 1024;
export const SECRET_VALUE_MAX_BYTES = 2 * 1024 * 1024;
const SECRET_REQUEST_MAX_BYTES = 64 * 1024;

export type SecretOperation = "keys" | "reveal" | "copy" | "update";

export interface SecretRouteTarget {
  clusterId: string;
  namespace: string;
  name: string;
  operation: SecretOperation;
}

type JsonObject = Record<string, unknown>;

function isRecord(value: unknown): value is JsonObject {
  return Boolean(value) && typeof value === "object" && !Array.isArray(value);
}

function decodePathPart(value: string, field: string): string {
  try {
    return decodeURIComponent(value);
  } catch {
    throw new RequestValidationError(
      400,
      "INVALID_IDENTIFIER",
      `${field} is not valid URL encoding`,
    );
  }
}

export function matchSecretRoute(
  method: string | undefined,
  pathname: string,
): SecretRouteTarget | null {
  const match = pathname.match(
    /^\/clusters\/([^/]+)\/secrets\/([^/]+)\/([^/]+)\/(keys|reveal|copy|update)$/,
  );
  if (!match) return null;

  const operation = match[4] as SecretOperation;
  if (
    (operation === "keys" && method !== "GET") ||
    (operation !== "keys" && method !== "POST")
  ) {
    return null;
  }

  return {
    clusterId: validateIdentifier(
      decodePathPart(match[1], "cluster_id"),
      "cluster_id",
      128,
    ),
    namespace: validateIdentifier(
      decodePathPart(match[2], "namespace"),
      "namespace",
    ),
    name: validateIdentifier(decodePathPart(match[3], "name"), "name"),
    operation,
  };
}

function revealTimeoutSeconds(configStore: ConfigStore): number {
  const configured = configStore.load().settings.secretRevealTimeoutSeconds;
  return Math.max(1, Math.min(300, Math.trunc(configured)));
}

export function secretDataMap(secret: JsonObject): Record<string, string> {
  const data = secret.data;
  if (!isRecord(data)) return {};

  const result: Record<string, string> = {};
  for (const [key, value] of Object.entries(data)) {
    result[String(key)] = String(value ?? "");
  }
  return result;
}

export function decodeBase64Strict(encoded: string): Buffer {
  if (encoded === "") return Buffer.alloc(0);

  const valid = /^(?:[A-Za-z0-9+/]{4})*(?:[A-Za-z0-9+/]{2}==|[A-Za-z0-9+/]{3}=)?$/.test(
    encoded,
  );
  if (!valid) {
    throw new Error("invalid base64");
  }

  const decoded = Buffer.from(encoded, "base64");
  if (decoded.toString("base64") !== encoded) {
    throw new Error("invalid base64");
  }
  return decoded;
}

export function isBinaryPayload(value: Buffer): boolean {
  if (value.length === 0) return false;
  if (value.includes(0)) return true;

  let textBytes = 0;
  for (const byte of value) {
    if (byte === 9 || byte === 10 || byte === 13 || (byte >= 32 && byte <= 126)) {
      textBytes += 1;
    }
  }

  return textBytes / value.length < 0.85;
}

export async function loadSecretRaw(
  target: SecretRouteTarget,
  configStore: ConfigStore,
  runner: KubectlRunner,
): Promise<JsonObject> {
  return runner.runJson(clusterCommand(
    configStore,
    target.clusterId,
    ["get", "secret", target.name, "-n", target.namespace, "-o", "json"],
    30,
    SECRET_JSON_MAX_OUTPUT_BYTES,
  ));
}

export function secretKeysPayload(
  secret: JsonObject,
  target: SecretRouteTarget,
  timeoutSeconds: number,
): JsonObject {
  const data = secretDataMap(secret);
  const keys = Object.keys(data).sort((left, right) => left.localeCompare(right)).map((key) => {
    const encoded = data[key] ?? "";
    let decodedBytes = 0;
    let validBase64 = true;
    let binary = false;

    try {
      const decoded = decodeBase64Strict(encoded);
      decodedBytes = decoded.length;
      binary = isBinaryPayload(decoded);
    } catch {
      validBase64 = false;
    }

    return {
      key,
      encodedBytes: Buffer.byteLength(encoded, "utf8"),
      decodedBytes,
      validBase64,
      binary,
    };
  });

  const metadata = isRecord(secret.metadata) ? secret.metadata : {};
  return {
    type: typeof secret.type === "string" && secret.type ? secret.type : "Opaque",
    immutable: Boolean(secret.immutable ?? false),
    namespace: typeof metadata.namespace === "string"
      ? metadata.namespace
      : target.namespace,
    name: typeof metadata.name === "string" ? metadata.name : target.name,
    keys,
    revealTimeoutSeconds: timeoutSeconds,
  };
}

async function readSecretKey(request: IncomingMessage): Promise<string> {
  const body = await readJsonBody(request, SECRET_REQUEST_MAX_BYTES);
  if (!isRecord(body) || typeof body.key !== "string") {
    throw new RequestValidationError(
      422,
      "INVALID_REQUEST",
      "Request body must contain a secret key",
    );
  }
  return validateIdentifier(body.key, "secret key", 512);
}

async function writeSecretKeys(
  response: ServerResponse,
  target: SecretRouteTarget,
  configStore: ConfigStore,
  runner: KubectlRunner,
): Promise<void> {
  const secret = await loadSecretRaw(target, configStore, runner);
  writeJson(
    response,
    secretKeysPayload(secret, target, revealTimeoutSeconds(configStore)),
  );
}

async function writeSecretReveal(
  request: IncomingMessage,
  response: ServerResponse,
  target: SecretRouteTarget,
  configStore: ConfigStore,
  auditStore: AuditStore,
  runner: KubectlRunner,
): Promise<void> {
  const key = await readSecretKey(request);
  const secret = await loadSecretRaw(target, configStore, runner);
  const data = secretDataMap(secret);

  if (!Object.prototype.hasOwnProperty.call(data, key)) {
    throw new RequestValidationError(
      404,
      "SECRET_KEY_NOT_FOUND",
      `Secret key was not found: ${key}`,
    );
  }

  let decoded: Buffer;
  try {
    decoded = decodeBase64Strict(data[key] ?? "");
  } catch {
    auditStore.append({
      action: "secret.reveal",
      status: "failed",
      clusterId: target.clusterId,
      namespace: target.namespace,
      resource: "secrets",
      name: target.name,
      message: "invalid base64 data",
      extra: { key },
    });
    throw new RequestValidationError(
      400,
      "SECRET_VALUE_INVALID_BASE64",
      `Secret key is not valid base64 data: ${key}`,
    );
  }

  if (decoded.length > SECRET_VALUE_MAX_BYTES) {
    auditStore.append({
      action: "secret.reveal",
      status: "failed",
      clusterId: target.clusterId,
      namespace: target.namespace,
      resource: "secrets",
      name: target.name,
      message: "secret value too large to reveal",
      extra: { key, decodedBytes: decoded.length },
    });
    throw new RequestValidationError(
      413,
      "SECRET_VALUE_TOO_LARGE",
      `Secret value is too large to reveal safely (${decoded.length} bytes)`,
    );
  }

  const binary = isBinaryPayload(decoded);
  auditStore.append({
    action: "secret.reveal",
    status: "success",
    clusterId: target.clusterId,
    namespace: target.namespace,
    resource: "secrets",
    name: target.name,
    extra: { key, decodedBytes: decoded.length, binary },
  });

  writeJson(response, {
    key,
    value: decoded.toString("utf8"),
    decodedBytes: decoded.length,
    binary,
    revealTimeoutSeconds: revealTimeoutSeconds(configStore),
  });
}

async function writeSecretCopy(
  request: IncomingMessage,
  response: ServerResponse,
  target: SecretRouteTarget,
  auditStore: AuditStore,
): Promise<void> {
  const key = await readSecretKey(request);
  auditStore.append({
    action: "secret.copy",
    status: "success",
    clusterId: target.clusterId,
    namespace: target.namespace,
    resource: "secrets",
    name: target.name,
    extra: { key },
  });
  writeJson(response, { ok: true });
}

async function writeSecretUpdate(request: IncomingMessage, response: ServerResponse, target: SecretRouteTarget, configStore: ConfigStore, auditStore: AuditStore, runner: KubectlRunner) {
  const body = await readJsonBody(request, SECRET_VALUE_MAX_BYTES + SECRET_REQUEST_MAX_BYTES);
  if (!isRecord(body) || typeof body.key !== "string" || typeof body.value !== "string") throw new RequestValidationError(422, "INVALID_REQUEST", "Request body must contain key and value");
  const key = validateIdentifier(body.key, "secret key", 512);
  const bytes = Buffer.byteLength(body.value, "utf8");
  if (bytes > SECRET_VALUE_MAX_BYTES) throw new RequestValidationError(413, "SECRET_VALUE_TOO_LARGE", "Secret value is too large");
  const secret = await loadSecretRaw(target, configStore, runner);
  if (secret.immutable === true) throw new RequestValidationError(409, "SECRET_IMMUTABLE", "Immutable Secret cannot be updated");
  const data = secretDataMap(secret);
  if (!Object.hasOwn(data, key)) throw new RequestValidationError(404, "SECRET_KEY_NOT_FOUND", "Secret key was not found");
  const metadata = isRecord(secret.metadata) ? secret.metadata : {};
  const resourceVersion = String(metadata.resourceVersion || "");
  const escape = (value: string) => value.replace(/~/g, "~0").replace(/\//g, "~1");
  const patch = JSON.stringify([{ op: "test", path: "/metadata/resourceVersion", value: resourceVersion }, { op: "replace", path: `/data/${escape(key)}`, value: Buffer.from(body.value, "utf8").toString("base64") }]);
  const command = clusterCommand(configStore, target.clusterId, ["patch", "secret", target.name, "-n", target.namespace, "--type=json", "--patch-file=-"], 30, SECRET_JSON_MAX_OUTPUT_BYTES);
  command.stdinText = patch;
  await runner.run(command);
  auditStore.append({ action: "secret.update", status: "success", clusterId: target.clusterId, namespace: target.namespace, resource: "secrets", name: target.name, extra: { key, decodedBytes: bytes } });
  writeJson(response, { ok: true });
}

function writeRouteError(
  response: ServerResponse,
  error: unknown,
  log: (message: string) => void,
): void {
  if (error instanceof RequestBodyError) {
    writeError(
      response,
      error.code === "REQUEST_TOO_LARGE" ? 413 : 400,
      error.code,
      error.message,
    );
    return;
  }
  if (error instanceof RequestValidationError) {
    writeError(response, error.statusCode, error.code, error.message);
    return;
  }
  if (error instanceof ClusterNotFoundError) {
    writeError(response, 404, "CLUSTER_NOT_FOUND", error.message);
    return;
  }
  if (error instanceof KubectlError) {
    writeKubectlError(response, error);
    return;
  }

  log("gateway secret operation failed");
  writeError(response, 500, "SECRET_OPERATION_FAILED", "Unable to process Secret operation");
}

export function handleSecretRequest(
  request: IncomingMessage,
  response: ServerResponse,
  pathname: string,
  configStore: ConfigStore,
  auditStore: AuditStore,
  runner: KubectlRunner,
  log: (message: string) => void,
): boolean {
  let target: SecretRouteTarget | null;
  try {
    target = matchSecretRoute(request.method, pathname);
  } catch (error) {
    writeRouteError(response, error, log);
    return true;
  }

  if (!target) return false;

  const operation = target.operation === "keys"
    ? writeSecretKeys(response, target, configStore, runner)
    : target.operation === "reveal"
      ? writeSecretReveal(
        request,
        response,
        target,
        configStore,
        auditStore,
        runner,
      )
      : target.operation === "copy"
        ? writeSecretCopy(request, response, target, auditStore)
        : writeSecretUpdate(request, response, target, configStore, auditStore, runner);

  void operation.catch((error) => writeRouteError(response, error, log));
  return true;
}
